File: tools/gnark/internal/abi/note_seizure_witness_binary.go
package abi

import (
	"bytes"
	"encoding/binary"
	"fmt"
)

const noteSeizureWitnessMagic = "NSZW"

type NoteSeizureRecoveryWitnessBinary struct {
	Commitment            [32]byte
	EPKAffine             PointAffineBinary
	C2                    [32]byte
	Salt                  [32]byte
	KeyConfirmation       [32]byte
	EncryptedAmount       [32]byte
	EncryptedNoteBlinding [32]byte
	Seed                  [32]byte
}

type NoteSeizureWitnessBinary struct {
	TotalLength                 uint32
	ClaimedStatementHash        [32]byte
	Anchor                      [32]byte
	NoteCommitment              [32]byte
	Nullifier                   [32]byte
	HistoryRequired             bool
	RecentPositionFloor         [32]byte
	AddressDiversifiedGenerator [32]byte
	AddressTransmissionKey      [32]byte
	AssetID                     [32]byte
	Amount                      [32]byte
	AuthorizationCommitment     [32]byte
	Recovery                    NoteSeizureRecoveryWitnessBinary
	NoteBlinding                [32]byte
	Position                    uint64
	StateCommitmentAuthPath     [][3][32]byte
	CNK                         [32]byte
	CnkCommitment               [32]byte
}

func EncodeNoteSeizureWitness(witness *NoteSeizureWitnessBinary) ([]byte, error) {
	var buf bytes.Buffer
	buf.WriteString(noteSeizureWitnessMagic)
	_ = binary.Write(&buf, binary.LittleEndian, uint32(0))
	for _, value := range [][32]byte{
		witness.ClaimedStatementHash,
		witness.Anchor,
		witness.NoteCommitment,
		witness.Nullifier,
	} {
		buf.Write(value[:])
	}
	if witness.HistoryRequired {
		buf.WriteByte(1)
	} else {
		buf.WriteByte(0)
	}
	for _, value := range [][32]byte{
		witness.RecentPositionFloor,
		witness.AddressDiversifiedGenerator,
		witness.AddressTransmissionKey,
		witness.AssetID,
		witness.Amount,
		witness.AuthorizationCommitment,
		witness.Recovery.Commitment,
	} {
		buf.Write(value[:])
	}
	for _, value := range [][32]byte{
		witness.Recovery.EPKAffine.X,
		witness.Recovery.EPKAffine.Y,
		witness.Recovery.C2,
		witness.Recovery.Salt,
		witness.Recovery.KeyConfirmation,
		witness.Recovery.EncryptedAmount,
		witness.Recovery.EncryptedNoteBlinding,
		witness.Recovery.Seed,
		witness.NoteBlinding,
	} {
		buf.Write(value[:])
	}
	_ = binary.Write(&buf, binary.LittleEndian, witness.Position)
	if len(witness.StateCommitmentAuthPath) != maxTriplePathLength {
		return nil, fmt.Errorf("triple path length %d must equal %d", len(witness.StateCommitmentAuthPath), maxTriplePathLength)
	}
	_ = binary.Write(&buf, binary.LittleEndian, uint32(len(witness.StateCommitmentAuthPath)))
	for _, siblings := range witness.StateCommitmentAuthPath {
		for _, sibling := range siblings {
			buf.Write(sibling[:])
		}
	}
	buf.Write(witness.CNK[:])
	buf.Write(witness.CnkCommitment[:])
	payload := buf.Bytes()
	if len(payload) > int(^uint32(0)) {
		return nil, fmt.Errorf("note seizure witness exceeds u32 length")
	}
	binary.LittleEndian.PutUint32(payload[4:8], uint32(len(payload)))
	return payload, nil
}

func DecodeNoteSeizureWitness(payload []byte) (*NoteSeizureWitnessBinary, error) {
	reader := bytes.NewReader(payload)
	magic, err := readExact(reader, 4)
	if err != nil {
		return nil, err
	}
	if string(magic) != noteSeizureWitnessMagic {
		return nil, fmt.Errorf("invalid note seizure witness magic %q", string(magic))
	}
	totalLength, err := readU32(reader)
	if err != nil {
		return nil, err
	}
	if totalLength != uint32(len(payload)) {
		return nil, fmt.Errorf("note seizure witness length mismatch: header=%d actual=%d", totalLength, len(payload))
	}
	witness := &NoteSeizureWitnessBinary{TotalLength: totalLength}
	for _, target := range []*[32]byte{
		&witness.ClaimedStatementHash,
		&witness.Anchor,
		&witness.NoteCommitment,
		&witness.Nullifier,
	} {
		if *target, err = read32(reader); err != nil {
			return nil, err
		}
	}
	if witness.HistoryRequired, err = readBool(reader); err != nil {
		return nil, err
	}
	for _, target := range []*[32]byte{
		&witness.RecentPositionFloor,
		&witness.AddressDiversifiedGenerator,
		&witness.AddressTransmissionKey,
		&witness.AssetID,
	} {
		if *target, err = read32(reader); err != nil {
			return nil, err
		}
	}
	if witness.Amount, err = readU128Field(reader); err != nil {
		return nil, err
	}
	if witness.AuthorizationCommitment, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.Recovery.Commitment, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.Recovery.EPKAffine, err = readPointAffine(reader); err != nil {
		return nil, err
	}
	for _, target := range []*[32]byte{
		&witness.Recovery.C2,
		&witness.Recovery.Salt,
		&witness.Recovery.KeyConfirmation,
		&witness.Recovery.EncryptedAmount,
		&witness.Recovery.EncryptedNoteBlinding,
		&witness.Recovery.Seed,
		&witness.NoteBlinding,
	} {
		if *target, err = read32(reader); err != nil {
			return nil, err
		}
	}
	if witness.Position, err = readU64(reader); err != nil {
		return nil, err
	}
	if witness.StateCommitmentAuthPath, err = readTriplePath(reader); err != nil {
		return nil, err
	}
	if witness.CNK, err = read32(reader); err != nil {
		return nil, err
	}
	if witness.CnkCommitment, err = read32(reader); err != nil {
		return nil, err
	}
	if reader.Len() != 0 {
		return nil, fmt.Errorf("note seizure witness has %d trailing bytes", reader.Len())
	}
	return witness, nil
}

File: crates/core/component/shielded-pool/src/gnark/note_seizure_witness_binary.rs
use anyhow::{bail, Context, Result};

use crate::gnark::{
    binary::{encode_triple_path_32, put_bytes, put_u32, put_u64, put_u8, BinaryCursor},
    note_seizure_witness::{NoteSeizureRecoveryWitness, NoteSeizureWitness},
    typed::encode_point_affine,
};

const NOTE_SEIZURE_WITNESS_MAGIC: &[u8; 4] = b"NSZW";

impl NoteSeizureWitness {
    pub fn encode(&self) -> Result<Vec<u8>> {
        let mut buf = Vec::new();
        put_bytes(&mut buf, NOTE_SEIZURE_WITNESS_MAGIC);
        put_u32(&mut buf, 0);
        for value in [
            &self.claimed_statement_hash,
            &self.anchor,
            &self.note_commitment,
            &self.nullifier,
        ] {
            put_bytes(&mut buf, value);
        }
        put_u8(&mut buf, u8::from(self.history_required));
        for value in [
            &self.recent_position_floor,
            &self.address_diversified_generator,
            &self.address_transmission_key,
            &self.asset_id,
            &self.amount,
            &self.authorization_commitment,
            &self.recovery.commitment,
        ] {
            put_bytes(&mut buf, value);
        }
        encode_point_affine(&mut buf, &self.recovery.epk_affine);
        for value in [
            &self.recovery.c2,
            &self.recovery.salt,
            &self.recovery.key_confirmation,
            &self.recovery.encrypted_amount,
            &self.recovery.encrypted_note_blinding,
            &self.recovery.seed,
            &self.note_blinding,
        ] {
            put_bytes(&mut buf, value);
        }
        put_u64(&mut buf, self.position);
        encode_triple_path_32(&mut buf, &self.state_commitment_auth_path)?;
        put_bytes(&mut buf, &self.rnk);
        put_bytes(&mut buf, &self.rnk_commitment);

        let total_length = u32::try_from(buf.len()).context("note seizure witness exceeds u32")?;
        buf[4..8].copy_from_slice(&total_length.to_le_bytes());
        Ok(buf)
    }

    pub fn decode(bytes: &[u8]) -> Result<Self> {
        let mut cursor = BinaryCursor::new(bytes);
        if cursor.read_fixed::<4>()? != *NOTE_SEIZURE_WITNESS_MAGIC {
            bail!("invalid note seizure witness magic")
        }
        let total_length = cursor.read_u32()?;
        ensure_length(total_length, bytes.len())?;
        let claimed_statement_hash = cursor.read_fixed::<32>()?;
        let anchor = cursor.read_fixed::<32>()?;
        let note_commitment = cursor.read_fixed::<32>()?;
        let nullifier = cursor.read_fixed::<32>()?;
        let history_required = cursor.read_bool()?;
        let recent_position_floor = cursor.read_fixed::<32>()?;
        let address_diversified_generator = cursor.read_fixed::<32>()?;
        let address_transmission_key = cursor.read_fixed::<32>()?;
        let asset_id = cursor.read_fixed::<32>()?;
        let amount = cursor.read_u128_field()?;
        let authorization_commitment = cursor.read_fixed::<32>()?;
        let recovery = NoteSeizureRecoveryWitness {
            commitment: cursor.read_fixed::<32>()?,
            epk_affine: cursor.read_point_affine()?,
            c2: cursor.read_fixed::<32>()?,
            salt: cursor.read_fixed::<32>()?,
            key_confirmation: cursor.read_fixed::<32>()?,
            encrypted_amount: cursor.read_fixed::<32>()?,
            encrypted_note_blinding: cursor.read_fixed::<32>()?,
            seed: cursor.read_fixed::<32>()?,
        };
        let note_blinding = cursor.read_fixed::<32>()?;
        let position = cursor.read_u64()?;
        let state_commitment_auth_path = cursor.read_triple_path_32()?;
        let rnk = cursor.read_fixed::<32>()?;
        let rnk_commitment = cursor.read_fixed::<32>()?;
        cursor.finish("note seizure witness")?;
        Ok(Self {
            total_length,
            claimed_statement_hash,
            anchor,
            note_commitment,
            nullifier,
            history_required,
            recent_position_floor,
            address_diversified_generator,
            address_transmission_key,
            asset_id,
            amount,
            authorization_commitment,
            recovery,
            note_blinding,
            position,
            state_commitment_auth_path,
            rnk,
            rnk_commitment,
        })
    }
}

fn ensure_length(encoded: u32, actual: usize) -> Result<()> {
    if encoded as usize != actual {
        bail!("note seizure witness length mismatch: header={encoded}, actual={actual}")
    }
    Ok(())
}
